proposals/draft/SIPXX - State Artifacts Composability.md

```diff
@@ -0,0 +1,250 @@
+# SIP x - Composable state artifacts
+
+## Status
+
+|  |  |
+| ------ | ------ |
+| State | Draft |
+| Issue Link | [#6](https://github.com/MeltanoLabs/Singer-Working-Group/issues/6) |
+| Discussion Thread(s) | (optional link) |
+| Created | 2021-10-25 |
+
+-----------------------
+
+## Proposal
+
+### TL;DR Overview
+
+Singer State artifacts can be split between its constituent streams and combined again.
+
+### What specific change do you propose to make?
+
+According to the [Singer specification](https://github.com/singer-io/getting-started/blob/master/docs/SPEC.md#state-message):
+
+> The semantics of a STATE value are not part of the specification, and should be determined independently by each Tap.
+
+This SIP proposes to standardize the structure of state message `value` objects in Singer taps, to allow splitting and combining state artifacts from multiple streams.
+
+The proposed `value` structure if a **flat** JSON object with the following schema:
+
+```json
+{
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "title": "Singer State",
+  "description": "State payload format for Singer taps",
+  "type": "object",
+  "properties": {
+    "bookmarks": {
+      "title": "Stream Bookmarks",
+      "type": "object",
+      "additionalProperties": false,
+      "patternProperties": {
+        "^\\w+(\\/\\w+\\=[A-Za-z0-9_-]+)*$": {
+          "type": "object",
+          "additionalProperties": true,
+          "properties": {
+            "replication_key": {
+              "type": "string"
+            },
+            "replication_key_value": {
+              "oneOf": [
+                {
+                  "type": "string"
+                },
+                {
+                  "type": "integer"
+                }
+              ]
+            }
+          }
+        }
+      }
+    },
+    "additionalProperties": false
+  }
+}
```

Comment:
@aaronsteers I've added a JSON schema for a **flat** state where partitions follow the Spark/Hive pattern `<table>/<key_1>=<value_1>/.../<key_n>=<value_n>`